### PR TITLE
Add generic hints to Cache

### PR DESCRIPTION
**Pull request by AJenbo**
This provides a way for IDEs and other analyzer tools to reason about the value returned by the cache and there by not cause lose of context when ever it is used. See https://github.com/laravel/framework/pull/38257 for a general discussions about this topic.

This PR is opened against master as generic annotations on the cache component is meant to come after 9.x.

P.s. @nunomaduro could you please share some more information regarding the plans for introducing generic annotations on the framework as it would make it easier to work on :)

### CODE REVIEW

**Comment by driesvints** [comment]
Can't this be sent to 9.x instead?

**Comment by AJenbo** [comment]
> Can't this be sent to 9.x instead?

I was told it coudn't

**Comment by driesvints** [comment]
Who did? @nunomaduro just said that we can't add these in general:

https://github.com/laravel/framework/pull/40759#issuecomment-1028237457

It's been over a year so maybe we can reconsider but it's up to @taylorotwell to decide.

**Comment by AJenbo** [comment]
Ok, I'm not privy to what the reasoning is. But I'm happy to have it go in 9.x if possible.

P.s. I'm pretty sure it's still 2022, or I forgot to shop up for xmas which would be pretty bad :D

**Comment by axlon** [comment]
These seem wrong to me, just because we pass a default doesn't mean that the value in the cache is of that type as well:

```
cache()->put('foo', 'bar');
cache()->get('foo', 123);
```

This will return a string but according to this PR it will return an int.

**Comment by taylorotwell** [comment]
Thanks for sending this in. I'm going to mark it as draft for now while people discuss it and also while we wait for @nunomaduro to return from vacation.

**Comment by AJenbo** [comment]
> These seem wrong to me, just because we pass a default doesn't mean that the value in the cache is of that type as well:
> 
> ```
> cache()->put('foo', 'bar');
> cache()->get('foo', 123);
> ```
> 
> This will return a string but according to this PR it will return an int.

I guess this is true, thought this feels like somewhat dubious usage.

The same issue could technically also occur with `remember()`:

```php
cache()->remember('foo', 60, function (): string { return 'bar'});
cache()->remember('foo', 60, function (): int { return 123});
```

For such cases I think it would be fair to still warn the user that they are likely doing something wrong.

File: src/Illuminate/Cache/Repository.php
Comment (nunomaduro) [inline src/Illuminate/Cache/Repository.php:180]: Please use `TCacheValue` everywhere, and remove `of mixed`.

File: src/Illuminate/Cache/Repository.php
Comment (nunomaduro) [inline src/Illuminate/Cache/Repository.php:184]: No need for `null` here.

File: src/Illuminate/Cache/Repository.php
Comment (nunomaduro) [inline src/Illuminate/Cache/Repository.php:183]: This should consideration that the `$default` value may be a closure that returns `TCacheValue`. And this comment can applied to other arguments annotations in this pull request.

File: types/Support/Cache.php
Comment (nunomaduro) [inline types/Support/Cache.php:1]: You will need to add tests for the cases when a closure is given for the $default arguments.

**Review by nunomaduro — CHANGES_REQUESTED** [review]
I've made some comments that must be addressed before merging. Also, can you rebase this on `v9.x` please?

**Comment by AJenbo** [comment]
> I've made some comments that must be addressed before merging. Also, can you rebase this on v9.x please?

Thanks! I'll look in to it :+1:

File: src/Illuminate/Cache/Repository.php
Comment (AJenbo) [reply src/Illuminate/Cache/Repository.php:180]: Oh neat I wasn't aware of that short hand

File: src/Illuminate/Cache/Repository.php
Comment (AJenbo) [reply src/Illuminate/Cache/Repository.php:183]: Wasn't aware of this possibility in Laravel, hopefully this will help more discover it as well :)

**Comment by AJenbo** [comment]
@nunomaduro the issues should now have been addressed and the PR has been rebased and targeted towards `9.x`

**Comment by AJenbo** [comment]
@nunomaduro the issues should now have been addressed and the PR has been rebased and targeted towards `9.x`

**Comment by driesvints** [comment]
@AJenbo can you maybe also fix the failing docblock check?

**Comment by driesvints** [comment]
Just copy paste the output of the build.

**Comment by AJenbo** [comment]
Well I have, and then added the following bit to make it actually valid:
```
 * @template TCacheValue
 *
```

@driesvints are you saying that we should go with the non functional generated one for now?

P.s. I think it would also be good to have the tool generate a diff to make the discrepancies more clear. But that is a bit out of scope here.

**Comment by driesvints** [comment]
@AJenbo just leave it for now then. We're automating this soon: https://github.com/laravel/framework/pull/45275

**Comment by timacdonald** [comment]
I'll fix up the generics Facade docblocks stuff once we merge this one. Don't let it be a blocker.

**Comment by nunomaduro** [comment]
Currently working on something else. But I will look at this week at this.

**Comment by nunomaduro** [comment]
Made a few adjustments, and added as you as co-author: https://github.com/laravel/framework/pull/45467.

**Comment by Tklaversma** [comment]
@nunomaduro Since this update, doing the following will give the error `'void' method 'rememberForever' result used`.

```php
    public function doStuff(): string
    {
        return Cache::rememberForever('key', function (): string {
            return 'example';
        });
    }
```

In other words, it sees the return type `TCacheValue` on `rememberForever()` as a void. Probably, because `TCacheValue` has no declaration.

**Comment by mvynhb** [comment]
Using template return types does not work for all IDE's intelephense! It might work for PHPStorm, but not for vscode and others. These changes were merged as if we all used PHPStorm...

**Pull request closed without merging**
